pkg/adminconfig/evaluator_interfaces.md

```diff
@@ -0,0 +1,259 @@
+// Used to evaluate configuration policies.
+type EvaluatorInput struct {
+	// Workload configuration
+	Workload WorkloadInfo `json:"workload"`
+	// Application properties
+	AppInfo api.ApplicationDetails `json:"application,omitempty"`
+	// Asset metadata
+	AssetMetadata *assetmetadata.DataDetails `json:"dataset"`
+	// Requirements for asset usage
+	AssetRequirements Request `json:"request"`
+	// Governance Actions for reading data (relevant for read scenarios only)
+	GovernanceActions []model.Action `json:"actions"`
+}
+```
+On top of the dynamic input based on a specific FybrikApplication, an `Infrastructure` object is used by the config policy evaluator to get deployment infrastructure information: available clusters, available storage, bandwidth metrics, etc. Evaluator uses `InfrastructureManager` to obtain the `Infrastructure` data.
+
+In the current implementation, `Infrastructure` is defined as:
+```
+type Infrastructure struct {
+	// Clusters available for deployment
+	Clusters []multicluster.Cluster `json:"clusters"`
+}
+```
+
+### Output (`adminconfig.EvaluatorOutput`)
+
+The result of evaluating config policies on a given dataset contains deployment `Decision` for each capablility, after merging all relevant decisions for the capability.
+ A conflict during evaluation result is marked by `Valid == false`
+ For each capability, the decision will provide the manager with the following information:
+ - whether the capability should be deployed, should not be deployed, or is allowed to be deployed based on manager decision.
+ - a restriction on the clusters where the capability will be deployed
+ - a restriction on the modules deploying this capability (e.g. deployment scope)
+ 
+```
+// EvaluatorOutput is an output of ConfigurationPoliciesEvaluator.
+// Used by manager to decide which modules are deployed and in which cluster.
+type EvaluatorOutput struct {
+	// Valid is true when there is no conflict between the decisions, and false otherwise
+	Valid bool
+	// Dataset identifier
+	DatasetID string
+	// Decisions per capability (after being merged)
+	ConfigDecisions DecisionPerCapabilityMap
+}
+```
+
+`Decision` is a result of evaluating a configuration policy which satisfies the predicates of the policy.
+`Deploy` is a deployment decision (require, forbid or allow).
+`DeploymentRestrictions` restricts the choice of the modules to be deployed, deployment clusters, storage accounts, etc.
+`Policy` combines IDs and descriptions of all policies that have been applied.
+
+```
+// DecisionPolicy is a justification for a policy that consists of a unique id and a human readable desciption
+type DecisionPolicy struct {
+	ID          string `json:"ID,omitempty"`
+	Description string `json:"description,omitempty"`
+}
+
+// Deployment restrictions on modules, clusters and additional resources
+type Restrictions struct {
+	// Restrictions on clusters used for deployment
+	Clusters []string `json:"clusters,omitempty"`
```

Comment:
Added a comment about adding restrictions on additional resources. However, it can not be absolutely generic as module attributes. For module attributes - just add some XXX field in a spec and use it in module restrictions without changing the code. Adding another concept as storage is more than that, manager needs to know how to deal with this restriction.